### PR TITLE
3D box surface plot

### PR DESCRIPTION
**Pull request by imogencresswell**
Plots a 3D visualisation of the sides and top of a cartesian simulation

### CODE REVIEW

File: plotpal/volumes.py [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+    side_info['z'] = np.where(side_bool, zz, np.nan)
+    side_info['surfacecolor'] = np.where(side_bool, data_vals, np.nan)
+
+    return side_info
+#turn colormesh into box with info on left side right side top, x, y and z basis, and colourmap
+class Box:
+    """
+    A struct containing information about a slice colormesh plot
+
+    # Attributes
```
Comment (evanhanders) [inline plotpal/volumes.py:104]: Update docstring with all relevant keyword arguments that are passed in on line 114.

File: plotpal/volumes.py [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+        x_basis, y_basis (strs) :
+            The dedalus basis names that the profile spans in the x- and y- direction of the plot
+        cmap  (str) :
+            The matplotlib colormap to plot the colormesh with
+        label (str):
+            A label for the colorbar
+
+    """
+
+    def __init__(self, left, right, top, x_basis='x', y_basis='y',z_basis='z', cmap='RdBu_r', \
```
Comment (evanhanders) [inline plotpal/volumes.py]: Reduce number of lines here to 2 or 3 (whatever looks nice and fits)

File: plotpal/volumes.py [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+                              pos_def=False, \
+                              vmin=None, vmax=None, log=False, vector_ind=None, \
+                              label=None, cmap_exclusion=0.005):
+        
+        self.first=True
```
Comment (evanhanders) [inline plotpal/volumes.py:121]: Add passed in attributes to self in the order that they are passed in

File: plotpal/volumes.py [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+        self.vector_ind = vector_ind
+        self.left=left
+        self.right=right
+        self.top=top
+
+
+        self.pos_def = pos_def
+        self.vmin = vmin
+        self.vmax = vmax
+        self.cmap_exclusion = cmap_exclusion
```
Comment (evanhanders) [inline plotpal/volumes.py]: Add a line here that if vmin != None and vmax != None, tell the user in a print statement that cmap_exclusion is being ignored.

File: plotpal/volumes.py [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+
+        self.cmap = cmap
+        self.label = label
+
+
+
+    def _modify_field(self, field):
+        #Subtract out m = 0
```
Comment (evanhanders) [inline plotpal/volumes.py]: delete line

File: plotpal/volumes.py [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+
+        if self.log: 
+            field = np.log10(np.abs(field))
+
+        return field
+
+    def _get_minmax(self, field):
+        # Get colormap bounds
+
```
Comment (evanhanders) [inline plotpal/volumes.py:162]: delete white space

File: plotpal/volumes.py [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+            vmin = -vmax
+
+        if self.vmin is not None:
+            vmin = self.vmin
+        if self.vmax is not None:
+            vmax = self.vmax
+
+        return vmin, vmax
+        
+#TODO update all following functions accordingly for plotly
```
Comment (evanhanders) [inline plotpal/volumes.py]: delete line

File: plotpal/volumes.py [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+        cb = matplotlib.colorbar.ColorbarBase(cax, cmap=cmap,
+                                norm=norm,
+                                orientation='horizontal')
+        cb.solids.set_rasterized(True)
+        cb.set_ticks((vmin, vmax))
+        cax.tick_params(direction='in', pad=1)
+        cb.set_ticklabels(('{:.2e}'.format(vmin), '{:.2e}'.format(vmax)))
+        cax.xaxis.set_ticks_position('bottom')
+        if self.label is not None:
+            #if self.vector_ind is not None:
```
Comment (evanhanders) [inline plotpal/volumes.py]: delete commented lines

File: plotpal/volumes.py [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+            z = d['z']
+            sfc = cmap(norm(d['surfacecolor']))
+            surf = ax.plot_surface(x, y, z, facecolors=sfc, cstride=1, rstride=1, linewidth=0, antialiased=False, shade=False)
+            ax.plot_wireframe(x, y, z, ccount=1, rcount=1, linewidth=1, color='black')
+        
+        
+
+        ax.view_init(azim=25, elev=10)
+
+
```
Comment (evanhanders) [inline plotpal/volumes.py]: delete white space and commented lines; add keyword args in __init__ to specify azim and elev, use these values as the defaults.

File: plotpal/volumes.py [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+        """
+        self.grid = None
+        super(BoxPlotter, self).__init__(*args, distribution='even-write', **kwargs)
+        self.counter = 0
+        self.boxes = []
+
+    def setup_grid(self, *args, **kwargs):
+        """ Initialize the plot grid for the colormeshes """
+        self.grid = RegularColorbarPlotGrid(*args, **kwargs, threeD=True)#copy past make subplots from ploty
+        #fig = go.Figure(layout={'width': 2000, 'height': 1000})
```
Comment (evanhanders) [inline plotpal/volumes.py]: delete commented lines

File: examples/d3/ivp_box_3d_mhd_shear/plot_box.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+if n_files is not None: 
+    n_files = int(n_files)
+
+# Create Plotter object, tell it which fields to plot
+plotter = BoxPlotter(root_dir, file_dir=data_dir, out_name=out_name, start_file=start_file, n_files=n_files)
+
+
+# Just plot a single plot (1x1 grid) of the field "b"
+plotter.setup_grid(num_rows=1, num_cols=1)
+plotter.add_box(left='w side',right='w front', top='w top', x_basis='x', y_basis='y',z_basis='z')
```
Comment (evanhanders) [inline examples/d3/ivp_box_3d_mhd_shear/plot_box.py]: add a label to the colormap here, and possibly add another column so it's fancier.

File: examples/d3/ivp_box_3d_mhd_shear/planar_Kolmogorov_3D.py [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+problem.substitutions['Ox'] = "(dy(w) - dz(v))"
+problem.substitutions['Oy'] = "(dz(u) - dx(w))"
+problem.substitutions['Oz'] = "(dx(v) - dy(u))"
+problem.substitutions['vol_avg(A)']   = 'integ(A)/Lx/Ly/Lz'
+
+# Note the pressure term in this formulation is really p + u^2/2
+problem.add_equation("dt(u) - Reinv*(dx(dx(u)) + dy(dy(u)) + dz(dz(u))) + dx(p) - MA2inv*Jy = v*Oz - w*Oy + MA2inv*(Jy*Bz - Jz*By)")
+problem.add_equation("dt(v) - Reinv*(dx(dx(v)) + dy(dy(v)) + dz(dz(v))) + dy(p) + MA2inv*Jx = w*Ox - u*Oz + MA2inv*(Jz*Bx - Jx*Bz)")
+problem.add_equation("dt(w) - Reinv*(dx(dx(w)) + dy(dy(w)) + dz(dz(w))) + dz(p) = u*Oy - v*Ox + MA2inv*(Jx*By - Jy*Bx) + sin(x)")
+# What's commented out here: old code where the momentum advection term was v dot grad v, as opposed to what's above
```
Comment (evanhanders) [inline examples/d3/ivp_box_3d_mhd_shear/planar_Kolmogorov_3D.py:85]: get rid of commented lines, and make sure that the example run at these parameters makes a semi-interesting plot :)

**Review by evanhanders — CHANGES_REQUESTED** [review]
Hey Immie! This is the first time I've done this side of a PR, let's see how this looks and please try to address these changes over the coming week or so.

**Comment by evanhanders** [comment]
@imogencresswell , also general rule of thumb for python is that functions should have 1 line of white space between them and classes should have two lines above them.

File: plotpal/volumes.py [hunk 13]
```diff
@@ -0,0 +1,373 @@
...
+
+    def __init__(self, *args, **kwargs):
+        """
+        Initializes the slice plotter.
+
+        # Arguments
+            *args, **kwargs : Additional keyword arguments for super().__init__() 
+        """
+        self.grid = None
+        self.cutout = False #this line says syntax error
```
Comment (imogencresswell) [inline plotpal/volumes.py]: syntax error here

File: plotpal/volumes.py [hunk 14]
```diff
@@ -0,0 +1,373 @@
...
+        self.grid = RegularColorbarPlotGrid(*args, **kwargs, threeD=True)
+
+
+    def add_box(self, *args, **kwargs):
+        self.boxes.append((self.counter, Box(*args, **kwargs)))
+        self.counter += 1
+    
+    def add_cutout_box(self, *args, **kwargs):
+        self.boxes.append((self.counter, Box(*args, **kwargs))
+        self.cutout = True #need an argument for if statement on line 382
```
Comment (imogencresswell) [inline plotpal/volumes.py]: I'm trying to get this passed through to line 350 (not 382)

File: plotpal/volumes.py [hunk 15]
```diff
@@ -0,0 +1,373 @@
...
+            axs, caxs = self._groom_grid()
+            tasks = []
+            for k, bx in self.boxes:
+                if bx.left not in tasks:
+                    tasks.append(bx.left)
+                if bx.right not in tasks:
+                    tasks.append(bx.right)
+                if bx.top not in tasks:
+                    tasks.append(bx.top)
+                if self.cutout is True:
```
Comment (imogencresswell) [inline plotpal/volumes.py]: so that I can add these so that you can plot a cut out box and a normal box if needed

File: plotpal/volumes.py [hunk 16]
```diff
@@ -0,0 +1,373 @@
...
+        side_list = [xy_side, xz_side, yz_side]
+        if left_mid is not None:
+            side_list = [xy_side, xz_side, yz_side, xy_mid, xz_mid, yz_mid]
+        for d in side_list:
+            x = d['x']
+            y = d['y']
+            z = d['z']
+            sfc = cmap(norm(d['surfacecolor']))
+            surf = ax.plot_surface(x, y, z, facecolors=sfc, cstride=1, rstride=1, linewidth=0, antialiased=False, shade=False)
+            ax.plot_wireframe(x, y, z, ccount=1, rcount=1, linewidth=1, color='black')
```
Comment (imogencresswell) [inline plotpal/volumes.py:288]: currently doesn't plot lines on the inside of the cutout

Comment (evanhanders) [reply plotpal/volumes.py:288]: TODO: (after our convo today) add wireframe outlines to the inside cutout. I think you only need to add to the vertical left and right slices (the horizontal slice will be filled in by all the other wireframes)

**Comment by evanhanders** [comment]
Merged in; I'll work on addressing the various comments that haven't been sorted over the next...whenever I get around to it. Thanks for getting this together @imogencresswell !

**Pull request merged**